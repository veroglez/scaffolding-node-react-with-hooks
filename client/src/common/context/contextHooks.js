import { func, shape } from 'prop-types';
import React, { createContext, useContext, useReducer } from 'react';

import l10n from '../l10n';

export const StateContext = createContext();

export const initialState = {
  characters: [],
};

export const globalReducer = (state, action) => {
  switch (action.type) {
    case 'update':
      return {
        ...state,
        ...action.payload,
      };
    default:
      return state;
  }
};

export const useStateValue = () => useContext(StateContext);

export const StateProvider = ({ reducer, initialState: state, children }) => (
  <StateContext.Provider value={useReducer(reducer, state)}>
    {children}
  </StateContext.Provider>
);

StateProvider.propTypes = {
  reducer: func,
  initialState: shape({}),
  children: shape({}),
};

StateProvider.defaultProps = {
  reducer: undefined,
  initialState,
  children: undefined,
};
