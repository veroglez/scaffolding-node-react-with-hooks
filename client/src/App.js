import React, { useEffect } from 'react';
import { useStateValue } from 'common/context/contextHooks';
import fetch from 'common/fetch';

const onFetch = async ({ dispatch }) => {
  const endpoint = 'http://localhost:3000';
  const service = 'characters';
  const characters = await fetch({ endpoint, service }).catch(err => err);

  dispatch({
    type: 'update',
    payload: {
      characters,
    },
  });
};


const App = () => {
  const [state, dispatch] = useStateValue();

  useEffect(() => { onFetch({ dispatch }); });

  return (
    <div className="App">
      holi
    </div>
  );
};

export default App;
