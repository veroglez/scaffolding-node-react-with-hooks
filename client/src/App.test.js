
import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as Context from 'common/context/contextHooks';
import App from './App';

Enzyme.configure({ adapter: new Adapter() });
global.fetch = require('jest-fetch-mock');

const STATE = {
  characters: [
    { name: 'Vero' },
    { name: 'Miguel' },
  ],
};

const HOC = () => (
  <Context.StateContext.Provider value={[STATE]}>
    <App />
  </Context.StateContext.Provider>
);

const flushPromises = async () => new Promise(resolve => setImmediate(resolve));
const dispatch = jest.fn();

describe('<App>', () => {
  // beforeEach(() => {
  //   fetch.resetMocks();
  // });

  it('render', () => {
    const tree = renderer.create(<HOC />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('dispath is called with', async () => {
    const stateUpdated = {
      type: 'update',
      payload: { ...STATE },
    };
    fetch.mockResponseOnce(JSON.stringify(STATE.characters));

    jest
      .spyOn(Context, 'useStateValue')
      .mockImplementation(() => [STATE, dispatch]);
    mount(<HOC />);

    await flushPromises();
    expect(dispatch).toHaveBeenCalledWith(stateUpdated);
  });
});
