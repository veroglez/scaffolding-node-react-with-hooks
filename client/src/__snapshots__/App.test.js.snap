// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App> render 1`] = `
<div
  className="App"
>
  holi
</div>
`;
